proof/refine/ARM/KHeap_R.thy

```diff
@@ -74,65 +74,187 @@ abbreviation (input)
 where
   "set_obj' ptr obj s \<equiv> set_ko' ptr (injectKO obj) s"
 
+(* FIXME: move *)
+(** reader monad **)
+
+(* if state s satisfies P, and if f s returns a value, the value satisfies Q *)
+abbreviation
+  ovalid :: "('s \<Rightarrow> bool) \<Rightarrow> ('s,'a) lookup \<Rightarrow> ('a \<Rightarrow> 's \<Rightarrow>  bool) \<Rightarrow> bool"
+  ("o\<lbrace>_\<rbrace>/ _ /\<lbrace>_\<rbrace>")
+where
+  "o\<lbrace>P\<rbrace> f \<lbrace>Q\<rbrace> \<equiv> \<forall>s. P s \<longrightarrow> (\<exists>y. f s = Some y) \<longrightarrow>  Q (the (f s)) s"
+
+abbreviation (input)
+  oinvariant :: "('s,'a) lookup \<Rightarrow> ('s \<Rightarrow> bool) \<Rightarrow> bool" ("_ o\<lbrace>_\<rbrace>" [59,0] 60)
+where
+  "oinvariant f P \<equiv> o\<lbrace>P\<rbrace> f \<lbrace>\<lambda>_. P\<rbrace>"
+
+lemma asks_wp[rule_format, rotated, simp]:
+  "o\<lbrace>\<lambda>s. P (f s) s\<rbrace> asks f \<lbrace>P\<rbrace>"
+  by(simp add:  split_def asks_def oreturn_def obind_def)
+
+(* no fail for reader monad *)
+definition
+  no_fail_r :: "('s \<Rightarrow> bool) \<Rightarrow> ('s, 'a) lookup \<Rightarrow> bool" where
+  "no_fail_r P m \<equiv> \<forall>s. P s \<longrightarrow> (\<exists>y. m s = Some y)"
+
+lemma no_fail_r_gets_the [wp, simp, intro!]:
+  "no_fail_r P f \<Longrightarrow> no_fail P (gets_the (f::'a kernel_r))"
+  by (wpsimp simp: no_fail_r_def)
+
+lemma no_fail_r_oassert_opt [simp, wp]:
+  "no_fail_r (\<lambda>_. P \<noteq> None) (oassert_opt P)"
+  by (simp add: no_fail_r_def oassert_opt_def split: option.splits)
+
+lemma no_fail__r_owhen [wp]:
+  "(P \<Longrightarrow> no_fail_r Q f) \<Longrightarrow> no_fail_r (if P then Q else \<top>) (owhen P f)"
+  by (simp add: no_fail_r_def owhen_def)
+
+lemma no_fail_r_ounless [wp]:
+  "(\<not>P \<Longrightarrow> no_fail_r Q f) \<Longrightarrow> no_fail_r (if P then \<top> else Q) (ounless P f)"
+  by (simp add: no_fail_r_def ounless_def)
+
+lemma no_fail_r_ofail [simp, wp]:
+  "no_fail_r \<bottom> ofail"
+  by (simp add: no_fail_r_def ofail_def)
+
+lemma non_fail_r_asks_simp[simp]:
+  "no_fail_r P (asks f)"
+  unfolding asks_def get_def oreturn_def obind_def no_fail_r_def
+  by simp
+
+lemma non_fail_r_asks[wp]:
+  "no_fail_r \<top> (asks f)"
+  by simp
```

Comment:
I just noticed that these go along with what was discussed between Matt and Gerwin about `P` vs `\<top>`.